### PR TITLE
refactor(forms): update required validator and checkbox validator to inherit abstractValidator

### LINKED ISSUE #42267: Regression introduced by new min max validators
# Bug Report

### Affected Package

@angular/forms 12+

### Is this a regression?

Yes, the previous version in which this bug was not present was: 11

### Description
@angular/forms 12 introduced min and max validators based on the min and max attributes presence in #39063.

Before that, the browser would react to min/max attributes by doing validation and also by preventing to introduce invalid value. Both of those **native features are now impossible to leverage**.

Also, before that, it was possible to deactivate the browser native features with a null value, such as `[min]="null"` or `[min]="undefined"`. This is now impossible to do in strict mode compilation, because the new directive only accept `string | number` and would **result in a compilation error**. Possible temporary workaround for that is to use `$any` casting similar to `[min]="configuration.min ?? $any(undefined)"`

image

### Suggested fixes

I would suggest that both directives create, and maintain, the equivalent attribute on the HTMLInputElement, so that browsers are able to react accordingly and adapt their UX.

And I would suggest to loosen the input type like so:

```diff
- @Input() min!: string|number;
+ @Input() min: string|number|null|undefined;
```

If those suggestions sound mergeable I could create PRs for it.

## Minimal Reproduction

- https://stackblitz.com/edit/min-max-regression-angular-11
- https://stackblitz.com/edit/min-max-regression-angular-12, same as above, but with strict mode disabled, and Angular 12

## Exception or Error

```
Type 'number | null' is not assignable to type 'string | number'.
```

## Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 12.0.1
Node: 14.17.0
Package Manager: yarn 1.22.5
OS: linux x64

Angular: 12.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, localize, material
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.1
@angular-devkit/build-angular   12.0.1
@angular-devkit/core            12.0.1
@angular-devkit/schematics      12.0.1
@angular/flex-layout            12.0.0-beta.34
@schematics/angular             12.0.1
ng-packagr                      12.0.0
rxjs                            6.6.7
typescript                      4.2.4

</code></pre>

### PR DESCRIPTION
**Pull request by iRealNirmal**
Modified required validator and checkbox validator to inherit abstractValidator.

For every validato type different PR will be raised as discussed in #42378.

Closes #42267

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #42378

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by iRealNirmal** [comment]
this PR is not ready for review, I will update once it can be reviewed.

**Comment by iRealNirmal** [comment]
@AndrewKushnir @crisbeto @devversion @petebacondarwin can you please share your input regarding failure of components-repo-unit-tests

File: packages/forms/src/directives/validators.ts [hunk 1]
```diff
@@ -347,41 +356,25 @@ export const CHECKBOX_REQUIRED_VALIDATOR: StaticProvider = {
...
+  override normalizeInput = (input: unknown): boolean => toBoolean(input);
 
-  /**
-   * Registers a callback function to call when the validator inputs change.
-   * @nodoc
-   */
-  registerOnValidatorChange(fn: () => void): void {
-    this._onChange = fn;
+  /** @internal */
+  override createValidator = (input: unknown): ValidatorFn => requiredValidator;
```
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/forms/src/directives/validators.ts:374]: Thanks for the investigation @devversion!

The error is legit and we should refactor the `CheckboxRequiredValidator` class as a part of this change as well, here is my proposal (omitting the `@Directive` part for brevity):

```
export class CheckboxRequiredValidator extends RequiredValidator {
  /** @nodoc */
  override createValidator = (input: unknown): ValidatorFn => requiredTrueValidator;
}
```

I also think that we should keep the getter/setter pair for the `required` input and do a coercion there for performance benefits: host bindings are being re-evaluated at each change detection cycle, so we'd be making this coercion every time (vs coercing the value only when it actually changes).

Also +1 to what Paul mentioned regarding the tests for the checkbox validators: let's make sure we have them and there is a sufficient coverage (which includes changing inputs and confirming the behavior.

[the remark this answers is not in the document]
Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts:374]: thanks @devversion @AndrewKushnir for your inputs, let me do changes. Thanks @petebacondarwin for your support as well on other issues where you helped a lot.

[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/forms/src/directives/validators.ts:374]: Yeah -  I like the idea of doing the `toBoolean` in the `required` setter - to save on doing it on every call to `ngOnChanges()`.

File: packages/forms/src/directives/validators.ts [hunk 2]
```diff
@@ -363,25 +370,20 @@ export class RequiredValidator implements Validator {
...
+  override inputName = 'required';
 
-  /**
-   * Registers a callback function to call when the validator inputs change.
-   * @nodoc
-   */
-  registerOnValidatorChange(fn: () => void): void {
-    this._onChange = fn;
+  /** @internal */
+  override normalizeInput = (input: unknown): boolean => toBoolean(input);
```
Comment (petebacondarwin) [inline packages/forms/src/directives/validators.ts:380]: The call to `toBoolean()` should no longer be needed here, right? Since you are now normalizing the value in the `required` setter.

Comment (petebacondarwin) [reply packages/forms/src/directives/validators.ts:380]: So this could be:

```suggestion
  override normalizeInput = input => input;
```

File: packages/forms/src/directives/validators.ts [hunk 3]
```diff
@@ -363,25 +370,20 @@ export class RequiredValidator implements Validator {
...
+
+  /** @internal */
+  override createValidator = (input: unknown): ValidatorFn => requiredValidator;
+
+  override enabled(): boolean {
+    return this.normalizeInput(this.required) as boolean;
```
Comment (petebacondarwin) [inline packages/forms/src/directives/validators.ts:386]: This should no longer need to call `this.normalizeInput()` since `required` is normalized in the setter.

File: packages/forms/src/directives/validators.ts [hunk 4]
```diff
@@ -347,12 +356,10 @@ export const CHECKBOX_REQUIRED_VALIDATOR: StaticProvider = {
   selector:
       ':not([type=checkbox])[required][formControlName],:not([type=checkbox])[required][formControl],:not([type=checkbox])[required][ngModel]',
   providers: [REQUIRED_VALIDATOR],
-  host: {'[attr.required]': 'required ? "" : null'}
+  host: {'[attr.required]': 'enabled() ? "" : null'}
```
Comment (petebacondarwin) [inline packages/forms/src/directives/validators.ts:359]: Since we are now normalizing the `required` property in its setter, we can revert this change back to just `required`, right?

File: packages/forms/src/directives/validators.ts [hunk 5]
```diff
@@ -411,17 +413,11 @@ export class RequiredValidator implements Validator {
   selector:
       'input[type=checkbox][required][formControlName],input[type=checkbox][required][formControl],input[type=checkbox][required][ngModel]',
   providers: [CHECKBOX_REQUIRED_VALIDATOR],
-  host: {'[attr.required]': 'required ? "" : null'}
+  host: {'[attr.required]': 'enabled() ? "" : null'}
```
Comment (petebacondarwin) [inline packages/forms/src/directives/validators.ts:416]: I think this can be reverted too.

Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts:416]: I wanted to keep it for uniformity, but if you and @AndrewKushnir both are on same page that we don't need it, I will revert it.

thanks @petebacondarwin for reviewing PR.

File: packages/forms/src/directives/validators.ts [hunk 6]
```diff
@@ -22,6 +22,15 @@ function toInteger(value: string|number): number {
   return typeof value === 'number' ? value : parseInt(value, 10);
 }
 
+/**
+ * Method that ensures that converts null, false or 'false' string to boolean.
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:26]:
```suggestion
 * Method that converts null, false or 'false' string to boolean.
```

File: packages/forms/src/directives/validators.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/forms/src/directives/validators.ts:380]: Right, the `normalizeInput` should be an identity function (as Pete have it in a diff).

File: packages/forms/src/directives/validators.ts (hunk 5 above)
Comment (AndrewKushnir) [reply packages/forms/src/directives/validators.ts:416]: I'd vote for having uniformity with other validators, so let's change it to `enabled()` (as we have it in the PR now).

File: packages/forms/src/directives/validators.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/forms/src/directives/validators.ts:386]: Right, I think this line should just directly reference the `required` field:
```suggestion
    return this.required;
```

Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts]: done but with minor change

File: packages/forms/src/directives/validators.ts (hunk 2 above)
Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts]: done but with minor difference

**Comment by iRealNirmal** [comment]
@AndrewKushnir @petebacondarwin I am  done with most code implementation, I will be adding test cases which is pending but mean while if you have any other input feel free to share.

Thanks.

File: packages/forms/src/directives/validators.ts [hunk 7]
```diff
@@ -363,25 +370,20 @@ export class RequiredValidator implements Validator {
...
-  registerOnValidatorChange(fn: () => void): void {
-    this._onChange = fn;
+  /** @internal */
+  override normalizeInput = (input: unknown) => input;
+
+  /** @internal */
+  override createValidator = (input: unknown): ValidatorFn => requiredValidator;
+
+  override enabled(): boolean {
+    return this.required as boolean;
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:386]: I believe we should not need the `as boolean` type cast here and instead we can change the `required` getter type to only `boolean` (since we do a conversion in the setter):
```
get required(): boolean { ... }
```

File: packages/forms/src/directives/validators.ts [hunk 8]
```diff
@@ -22,6 +22,15 @@ function toInteger(value: string|number): number {
...
 
+/**
+ * Method that converts null, false or 'false' string to boolean.
+ *
+ * @param value input value
+ * @returns value of parameter in boolean.
+ */
+function toBoolean(input: unknown): boolean {
+  return input != null && input !== false && `${input}` !== 'false';
+}
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:33]: Minor note: could you please add a newline after the function (to visually separate it from the next comment)?

**Review by AndrewKushnir — COMMENTED** [review]
@iRealNirmal thanks for additional changes, I've just left a couple comments.

**Comment by iRealNirmal** [comment]
@AndrewKushnir @petebacondarwin I  was going through test cases for checkbox but I found we already have test cases for
`required validator should validate checkbox`.

It's already covering test case of  check and uncheck. Please let me know in case anything else needs to be done.

Thanks.

**Comment by AndrewKushnir** [comment]
@iRealNirmal thanks for checking this. It looks like the `required validator should validate checkbox` test in the `packages/forms/test/template_integration_spec.ts` file does most of the testing already, as you pointed out. I'd propose adding a check (to the same test) where we change the `required` directive input to `false` (`fixture.componentInstance.required = false;`) and make sure that the behavior is correct (that the validator is dynamically disabled).

**Comment by iRealNirmal** [comment]
@AndrewKushnir I have added some test cases let me know if you have any other test case scenarios as well which you want me to add it

File: goldens/public-api/forms/forms.md [hunk 9]
```diff
@@ -683,11 +684,11 @@ export class ReactiveFormsModule {
 }
 
 // @public
-export class RequiredValidator implements Validator {
-    registerOnValidatorChange(fn: () => void): void;
-    get required(): boolean | string;
+export class RequiredValidator extends AbstractValidatorDirective {
+    // (undocumented)
+    enabled(): boolean;
+    get required(): boolean;
```
Comment (AndrewKushnir) [inline goldens/public-api/forms/forms.md:690]: Note: this might be a concern from `public-api` perspective as we are narrowing the type. In practice that might become an issue in case the `RequiredValidator` is extended, but this is something that Angular does not guarantee to work correctly (with some exceptions when we specifically mark classes as suitable for extension). Let's keep things as is for now though and get more feedback from other reviewers.

File: packages/core/test/bundling/forms_template_driven/bundle.golden_symbols.json [hunk 10]
```diff
@@ -1412,6 +1418,9 @@
   {
     "name": "renderView"
   },
+  {
+    "name": "requiredValidator"
```
Comment (AndrewKushnir) [inline packages/core/test/bundling/forms_template_driven/bundle.golden_symbols.json:1422]: It's unclear why this symbol was not present before, since it was used in the previous implementation too. No actions needed, just some extra context for other reviewers.

File: packages/forms/src/directives/validators.ts [hunk 11]
```diff
@@ -22,6 +22,16 @@ function toInteger(value: string|number): number {
   return typeof value === 'number' ? value : parseInt(value, 10);
 }
 
+/**
+ * Method that converts null, false or 'false' string to boolean.
+ *
+ * @param value input value
+ * @returns value of parameter in boolean.
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:29]:
```suggestion
 * @param value input value.
 * @returns value of parameter converted to boolean.
```

File: goldens/public-api/forms/forms.md (hunk 9 above)
Comment (AndrewKushnir) [reply goldens/public-api/forms/forms.md:690]: In fact we always returned boolean, so it's mainly a type-only change, which we can roll back to `string|boolean` if that becomes a concern.

**Review by AndrewKushnir — COMMENTED** [review]
@iRealNirmal thanks for adding extra tests.
The change looks good (just 1 small comment on the docs), could you please change PR status from Draft -> Ready for Review if there are no other major changes that you plan to make?

File: packages/forms/src/directives/validators.ts (hunk 11 above)
Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts]: okay and to keep uniformity I have updated text to other 2 methods as well

**Review by AndrewKushnir — APPROVED** [review]
@iRealNirmal thanks for addressing the feedback, the changes look good 👍 

Note: there is a [legit CI failure](https://app.circleci.com/pipelines/github/angular/angular/40309/workflows/67ed31f7-2108-4217-becb-b028bc2337c1/jobs/1092248) related to the payload size for the Forms app:

```
FAIL: Commit undefined uncompressed main exceeded expected size
by 500 bytes or >1% (expected: 162474, actual: 163017).
```

This message is correct and there is indeed an increase for the `forms` test app. The reason is that [only "required" validator is used in the template](https://github.com/angular/angular/blob/master/integration/forms/src/app/app.component.ts#L11). Previously, the implementation of the required validator was more lightweight, but there was no code reuse across other validator directives. Now with the code reuse the size of the required validator has increased (due to the AbstractValidatorDirective used as a base class), but this starts to pay off once there are other validators in the template (of a given component or in other parts of an app), which is much more common. My proposal is to update the golden file (the `goldens/size-tracking/integration-payloads.json` one) with the new payload size.

File: goldens/public-api/forms/forms.md [hunk 12]
```diff
@@ -164,7 +164,8 @@ export class CheckboxControlValueAccessor extends BuiltInControlValueAccessor im
 
 // @public
 export class CheckboxRequiredValidator extends RequiredValidator {
-    validate(control: AbstractControl): ValidationErrors | null;
+    // (undocumented)
```
Comment (dylhunn) [inline goldens/public-api/forms/forms.md:167]: Looks like this method needs a JSDoc comment.

File: goldens/public-api/forms/forms.md (hunk 9 above)
Comment (dylhunn) [reply goldens/public-api/forms/forms.md:690]: I think we should go ahead and narrow it. If there is any breakage, we can fix forward very easily.

**Review by dylhunn — APPROVED** [review]
Looks good, thanks!

reviewed-for: fw-forms

**Review by dylhunn — APPROVED** [review]
reviewed-for: fw-core, fw-forms, public-api, size-tracking

File: goldens/public-api/forms/forms.md (hunk 12 above)
Comment (AndrewKushnir) [reply goldens/public-api/forms/forms.md:167]: Great catch @dylhunn 👍 We should not have this function in the public API. It looks look the issue is that we have `@nodoc` annotation on it instead of `@internal`. @iRealNirmal could you please update it?

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: public-api, fw-core

File: goldens/public-api/forms/forms.md (hunk 12 above)
Comment (iRealNirmal) [reply goldens/public-api/forms/forms.md:167]: updated thanks

**Comment by AndrewKushnir** [comment]
@iRealNirmal thanks for the update! FYI I've pushed a fixup commit to add the `/** @nodoc */` annotation for the `enabled` method of the `RequiredValidator` class (so that the method doesn't show up in docs). I've also rebased the PR on top of the latest master, since the payload size has changed there, which would probably avoid the need to update it in this PR (if that'd be required, I can push one more fixup commit).

Thanks for your work on this PR @iRealNirmal! We'll perform final reviews, run tests in Google's codebase and let you know if any additional changes are required.

File: goldens/size-tracking/integration-payloads.json [hunk 13]
```diff
@@ -52,7 +52,7 @@
     "master": {
       "uncompressed": {
         "runtime": 1063,
-        "main": 162474,
+        "main": 163160,
```
Comment (AndrewKushnir) [inline goldens/size-tracking/integration-payloads.json:55]: This payload size increase is legit. The reason is that [only "required" validator is used in the test app template](https://github.com/angular/angular/blob/master/integration/forms/src/app/app.component.ts#L11). Previously, the implementation of the required validator was more lightweight, but there was no code reuse across other validator directives. Now with the code reuse the size of the required validator has slightly increased (due to the `AbstractValidatorDirective` used as a base class), but this starts to pay off once there are other validators in the template (of a given component or in other parts of an app), which is much more common than having only a single "required" validator.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking, public-api

**Comment by AndrewKushnir** [comment]
FYI, we've came across an issue with property renaming and there is PR #44500 in progress to resolve that. This PR would require some updates once PR #44500 lands (I can push a fixup commit), so I'm marking it as blocked for now.

File: packages/forms/src/directives/validators.ts [hunk 14]
```diff
@@ -347,41 +358,35 @@ export const CHECKBOX_REQUIRED_VALIDATOR: StaticProvider = {
   selector:
       ':not([type=checkbox])[required][formControlName],:not([type=checkbox])[required][formControl],:not([type=checkbox])[required][ngModel]',
   providers: [REQUIRED_VALIDATOR],
-  host: {'[attr.required]': 'required ? "" : null'}
+  host: {'[attr.required]': 'enabled() ? "" : null'}
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:361]: @iRealNirmal since PR #44500 has landed, this PR is now unblocked.

Similar to my comment in https://github.com/angular/angular/pull/44545#discussion_r779783401, we can simplify the logic in this validator and it can look like this:
```
@Directive({
  selector:
      ':not([type=checkbox])[required][formControlName],:not([type=checkbox])[required][formControl],:not([type=checkbox])[required][ngModel]',
  providers: [REQUIRED_VALIDATOR],
  host: {'[attr.required]': '_enabled ? "" : null'}
export class RequiredValidator extends AbstractValidatorDirective {
  /**
   * @description
   * Tracks changes to the required attribute bound to this directive.
   */
  @Input()
  required!: boolean|string;

  /** @internal */
  override inputName = 'required';

  /** @internal */
  override normalizeInput = (input: unknown): boolean => toBoolean(input);

  /** @internal */
  override createValidator = (input: boolean): ValidatorFn => requiredValidator;

  /** @nodoc */
  override enabled(input: boolean): boolean {
    return input;
  }
}
```

We don't need extra caching of the enabled state, it's now a part of the base class.

Comment (AndrewKushnir) [reply packages/forms/src/directives/validators.ts:361]: I believe this would also allow us to avoid payload size increase, let's try to revert the change in `goldens/size-tracking/integration-payloads.json` and see what the CI status is.

Comment (iRealNirmal) [reply packages/forms/src/directives/validators.ts:370]: sorry was working on other open PRs, I have rebased branch will try to close this ticket in a day or two

**Comment by iRealNirmal** [comment]
@AndrewKushnir seems it's still failing for payload.json can you provide your input here ?

**Comment by AndrewKushnir** [comment]
> seems it's still failing for payload.json can you provide your input here ?

@iRealNirmal I think we can proceed with the payload size update in the golden file (similar to what you had before in this PR).

File: packages/forms/src/directives/validators.ts [hunk 15]
```diff
@@ -420,17 +417,11 @@ export class RequiredValidator implements Validator {
   selector:
       'input[type=checkbox][required][formControlName],input[type=checkbox][required][formControl],input[type=checkbox][required][ngModel]',
   providers: [CHECKBOX_REQUIRED_VALIDATOR],
-  host: {'[attr.required]': 'required ? "" : null'}
+  host: {'[attr.required]': 'enabled() ? "" : null'}
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:420]: Similar to the `RequiredValidator`, we should use `_enabled` here (instead of `enabled()`):
```suggestion
  host: {'[attr.required]': '_enabled ? "" : null'}
```

**Review by AndrewKushnir — COMMENTED** [review]
@iRealNirmal thanks for additional updates, I've just left a comment on the host bindings in the `CheckboxRequiredValidator` class. Could you please take a look when you get a chance?

**Comment by AndrewKushnir** [comment]
[Presubmit](https://test.corp.google.com/OCL:423594634:BASE:423594768:1642917097493:a1e84e0d).

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 531d1cf9a3c034380bbdfe803606ff1c6e04eb8a.

**Comment by AndrewKushnir** [comment]
@iRealNirmal thanks for all your help with the refactoring! 👍 This PR is now merged.

It looks like the last remaining validator class that require similar refactoring is `PatternValidator`. Please let me know if you might be interested in helping with that.

**Comment by iRealNirmal** [comment]
Yes @AndrewKushnir I will work on in this week once i get my spare time
